src/main.ts

```diff
@@ -179,8 +170,71 @@ export async function getOrCreateNftCollections(ctx: Context, cache: Cache, coll
             blockchain: collectionData.blockchain,
             contractType: collectionData.contractType
        });
+       await fillCollectionMetadata(ctx, latestBlockNumber, nftCollenctionEntity)
        cache.NftCollections.set(collectionId, nftCollenctionEntity);
     }
 }
 
+export async function fillTokenMetadata(ctx: Context, latestBlockNumber: number, token: NftEntity){
+    let tokenUri = null
+    if(token.nftCollection.baseUri){
+        tokenUri = token.nftCollection.baseUri.replace('{id}', token.tokenId.toString())
+    } else{
+        
+        try{
+            const contract = new erc721.Contract(ctx, {height: latestBlockNumber}, token.nftCollection.address);
+            tokenUri = await contract.tokenURI(token.tokenId)
+            if(tokenUri.includes('{id}')){
+                token.nftCollection.baseUri = tokenUri
+                tokenUri = tokenUri.replace('{id}', token.tokenId.toString())
+            }
+        } catch (e){
+            ctx.log.warn(`Failed to get tokenURI for token ${token.id}  ${(e as Error).message}`)
+            return 
+        }
+    }
+    token.uri = tokenUri;
+    const tokenMetadata = await parseTokenMetadata(ctx, tokenUri);
+    token.name = tokenMetadata?.name,
+    token.attributes = tokenMetadata?.attributes
+    token.description = tokenMetadata?.description
+    token.externalUrl = tokenMetadata?.external_url
+    token.image = tokenMetadata?.image
+    token.animationUrl = tokenMetadata?.animation_url
+
+}
+
+export async function fillCollectionMetadata(ctx: Context, latestBlockNumber: number, collection: NftCollectionEntity){
+    const contract = new erc721.Contract(ctx, {height: latestBlockNumber}, collection.address);
+    let contractUri = null
+    try{
+        contractUri = await contract.contractURI();
+    } catch{
+        contractUri = null
+    }
+    if (contractUri !== null){
+        collection.uri = contractUri
+        Object.assign(collection, await parseContractMetadata(ctx, contractUri))
+        
+    } else {
+        collection.name = await contract.name()
```

Comment:
Cant we access all these async data in one call?

Reply:
No, but for now, i changed it to make this calls at least asynchronously